### PR TITLE
[Serializer] Remove BC layer

### PR DESCRIPTION
**Pull request by lyrixx**
| Q             | A
| ------------- | ---
| Branch?       | 7.0
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       |
| License       | MIT
| Doc PR        |

### CODE REVIEW

**Review by wouterj — CHANGES_REQUESTED** [review]
Thanks for working on this!

Blocking as the necessary CHANGELOG and UPGRADE guide changes are missing ;)

**Comment by lyrixx** [comment]
> Blocking as the necessary CHANGELOG and UPGRADE guide changes are missing ;)

Yeah 👍🏼 But I don't remember what are the new guidelines for 7.0. Is there a summary of what I must do for each files?

**Comment by tucksaun** [comment]
It was on my todo list for the summer, thank you for taking care of it 😌

**Comment by GromNaN** [comment]
> Is there a summary of what I must do for each files?

[UPGRADE-7.0.md](https://github.com/symfony/symfony/blob/6.4/UPGRADE-7.0.md) needs to be updated on the branch 7.0 for each deprecated feature that is removed. As [you tried before](https://github.com/symfony/symfony/commit/f8730746fde29a02582f4d81b8ec27cf91fc8fd4#diff-6d49df895edd288cee25c95df434ae329795980f8024cdeb33ccd66ab65fb52c).

**Comment by nicolas-grekas** [comment]
See fabbot also (except the false-positive on the switch)

File: src/Symfony/Component/Serializer/Normalizer/ConstraintViolationListNormalizer.php [hunk 1]
```diff
@@ -41,7 +41,7 @@ public function __construct(
     public function getSupportedTypes(?string $format): array
     {
         return [
-            ConstraintViolationListInterface::class => __CLASS__ === static::class || $this->hasCacheableSupportsMethod(),
+            ConstraintViolationListInterface::class => __CLASS__ === static::class,
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/ConstraintViolationListNormalizer.php]: true, since it's final now (same for every other similar case)

**Review by nicolas-grekas — COMMENTED** [review]
Please add lines to the changelog of the component, same to the root UPGRADE-7.0.md file, and possibly ake the title of this PR more useful for a changelog

File: src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php [hunk 2]
```diff
@@ -42,35 +42,19 @@ class GetSetMethodNormalizer extends AbstractObjectNormalizer
...
+    public function supportsNormalization(mixed $data, string $format = null, array $context = []): bool
     {
         return parent::supportsNormalization($data, $format) && $this->supports($data::class);
     }
 
-    /**
-     * @param array $context
-     */
-    public function supportsDenormalization(mixed $data, string $type, string $format = null /* , array $context = [] */): bool
+    public function supportsDenormalization(mixed $data, string $type, string $format = null, array $context = []): bool
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php:53]: missing changelog/upgrade lines for these :)

Comment (lyrixx) [reply src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php:53]: Done, but only for the related interaface

**Comment by lyrixx** [comment]
I don't know how to fix theses tests in https://github.com/symfony/symfony/actions/runs/5406243227/jobs/9822895029?pr=50736#step:8:9587


```
1) Symfony\Component\Messenger\Tests\Stamp\ErrorDetailsStampTest::testDeserialization
Error: Interface "Symfony\Component\Serializer\Normalizer\ContextAwareNormalizerInterface" not found

/home/runner/work/symfony/symfony/src/Symfony/Component/Messenger/Transport/Serialization/Normalizer/FlattenExceptionNormalizer.php:25
/home/runner/work/symfony/symfony/src/Symfony/Component/Messenger/vendor/symfony/error-handler/DebugClassLoader.php:296
/home/runner/work/symfony/symfony/src/Symfony/Component/Messenger/Tests/Stamp/ErrorDetailsStampTest.php:62
```

**Comment by nicolas-grekas** [comment]
> fix theses tests

Can we stop using `ContextAwareNormalizerInterface` in Messenger on branch 6.4? If not, we'll have to declare on branch 6.4 that Messenger 6.4 doesn't work with serializer ^7

**Comment by lyrixx** [comment]
I checked, and We don't allow 7 in lower branches 
https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/ErrorHandler/composer.json#L25

> Can we stop using `ContextAwareNormalizerInterface` in Messenger on branch 6.4?

I can check

**Comment by derrabus** [comment]
> Can we stop using `ContextAwareNormalizerInterface` in Messenger on branch 6.4?

We probably can by dropping support for Serializer 5 which would be acceptable imho.

**Comment by lyrixx** [comment]
> Can we stop using `ContextAwareNormalizerInterface` in Messenger on branch 6.4

Done ➡ https://github.com/symfony/symfony/pull/50824

File: src/Symfony/Component/Serializer/Tests/Normalizer/TestNormalizer.php [hunk 3]
```diff
@@ -25,6 +25,9 @@ public function normalize($object, string $format = null, array $context = []):
         return null;
     }
 
+    /**
+     * @return array<class-string|'*'|'object'|string, bool|null>
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Tests/Normalizer/TestNormalizer.php]: those should all be removed, inheritance FTW

File: UPGRADE-7.0.md [hunk 4]
```diff
@@ -47,3 +47,9 @@ Serializer
 
  * Remove denormalization support for `AbstractUid` in `UidNormalizer`, use one of `AbstractUid` child class instead
  * Denormalizing to an abstract class in `UidNormalizer` now throws an `\Error`
+ * Remove `ContextAwareDenormalizerInterface`, use `DenormalizerInterface` instead
+ * Remove `ContextAwareNormalizerInterface`, use `NormalizerInterface` instead
+ * Remove `CacheableSupportsMethodInterface`, use `NormalizerInterface` and `DenormalizerInterface` instead
+ * First argument of `Symfony\Component\Validator\Mapping\ClassMetadata::setSerializedName()` is now required
+ * Third argument `array $context = []` of the `Symfony\Component\Serializer\Normalizer\NormalizerInterface::supportsNormalization()` is now required
+ * Forth argument `array $context = []` of the `Symfony\Component\Serializer\Normalizer\DenormalizerInterface::supportsDenormalization()` is now required
```
Comment (nicolas-grekas) [inline UPGRADE-7.0.md]:
```suggestion
 * Fourth argument `array $context = []` of the `DenormalizerInterface::supportsDenormalization()` is now required
```

File: UPGRADE-7.0.md [hunk 5]
```diff
@@ -47,3 +47,9 @@ Serializer
 
  * Remove denormalization support for `AbstractUid` in `UidNormalizer`, use one of `AbstractUid` child class instead
  * Denormalizing to an abstract class in `UidNormalizer` now throws an `\Error`
+ * Remove `ContextAwareDenormalizerInterface`, use `DenormalizerInterface` instead
+ * Remove `ContextAwareNormalizerInterface`, use `NormalizerInterface` instead
+ * Remove `CacheableSupportsMethodInterface`, use `NormalizerInterface` and `DenormalizerInterface` instead
+ * First argument of `Symfony\Component\Validator\Mapping\ClassMetadata::setSerializedName()` is now required
+ * Third argument `array $context = []` of the `Symfony\Component\Serializer\Normalizer\NormalizerInterface::supportsNormalization()` is now required
```
Comment (nicolas-grekas) [inline UPGRADE-7.0.md]:
```suggestion
 * First argument of `ClassMetadata::setSerializedName()` is now required
 * Third argument `array $context = []` of the `NormalizerInterface::supportsNormalization()` is now required
```

File: src/Symfony/Component/Serializer/CHANGELOG.md [hunk 6]
```diff
@@ -6,6 +6,12 @@ CHANGELOG
 
  * Remove denormalization support for `AbstractUid` in `UidNormalizer`, use one of `AbstractUid` child class instead
  * Denormalizing to an abstract class in `UidNormalizer` now throws an `\Error`
+ * Remove `ContextAwareDenormalizerInterface`, use `DenormalizerInterface` instead
+ * Remove `ContextAwareNormalizerInterface`, use `NormalizerInterface` instead
+ * Remove `CacheableSupportsMethodInterface`, use `NormalizerInterface` and `DenormalizerInterface` instead
+ * First argument of `Symfony\Component\Validator\Mapping\ClassMetadata::setSerializedName()` is now required
+ * Third argument `array $context = []` of the `Symfony\Component\Serializer\Normalizer\NormalizerInterface::supportsNormalization()` is now required
+ * Forth argument `array $context = []` of the `Symfony\Component\Serializer\Normalizer\DenormalizerInterface::supportsDenormalization()` is now required
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/CHANGELOG.md]:
```suggestion
 * First argument of `ClassMetadata::setSerializedName()` is now required
 * Third argument `array $context = []` of the `NormalizerInterface::supportsNormalization()` is now required
 * Fourth argument `array $context = []` of the `DenormalizerInterface::supportsDenormalization()` is now required
```

File: src/Symfony/Component/Serializer/Normalizer/DataUriNormalizer.php [hunk 7]
```diff
@@ -46,7 +46,7 @@ public function __construct(MimeTypeGuesserInterface $mimeTypeGuesser = null)
 
     public function getSupportedTypes(?string $format): array
     {
-        $isCacheable = __CLASS__ === static::class || $this->hasCacheableSupportsMethod();
+        $isCacheable = true;
 
         return [
             \SplFileInfo::class => $isCacheable,
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/DataUriNormalizer.php]:
```suggestion
        return [
            \SplFileInfo::class => true,
```

and the next lines also of course

**Review by nicolas-grekas — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @lyrixx.

**Pull request merged**